File: packages/services/octobot_services/interfaces/util/portfolio.py
#  Drakkar-Software OctoBot-Services
#  Copyright (c) Drakkar-Software, All rights reserved.
#
#  This library is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public
#  License as published by the Free Software Foundation; either
#  version 3.0 of the License, or (at your option) any later version.
#
#  This library is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#
#  You should have received a copy of the GNU Lesser General Public
#  License along with this library.
import decimal

import octobot_commons.constants as constants

import octobot_services.interfaces as interfaces
import octobot_trading
import octobot_trading.api as trading_api


def _merge_portfolio_values(portfolio1, portfolio2):
    for key, value in portfolio2.items():
        if key in portfolio1:
            portfolio1[key] += value
        else:
            portfolio1[key] = value
    return portfolio1


def get_portfolio_holdings():
    real_currency_portfolio = {}
    simulated_currency_portfolio = {}

    for exchange_manager in interfaces.get_exchange_managers():
        if trading_api.is_trader_existing_and_enabled(exchange_manager):

            trader_currencies_values = trading_api.get_current_holdings_values(exchange_manager)
            if trading_api.is_trader_simulated(exchange_manager):
                _merge_portfolio_values(simulated_currency_portfolio, trader_currencies_values)
            else:
                _merge_portfolio_values(real_currency_portfolio, trader_currencies_values)
    return real_currency_portfolio, simulated_currency_portfolio


def get_portfolio_current_value():
    simulated_value = 0
    real_value = 0
    has_real_trader = False
    has_simulated_trader = False

    for exchange_manager in interfaces.get_exchange_managers():
        if trading_api.is_trader_existing_and_enabled(exchange_manager):

            current_value = trading_api.get_current_portfolio_value(exchange_manager)

            # current_value might be 0 if no trades have been made / canceled => use origin value
            if current_value == 0:
                current_value = trading_api.get_origin_portfolio_value(exchange_manager)

            if trading_api.is_trader_simulated(exchange_manager):
                simulated_value += current_value
                has_simulated_trader = True
            else:
                real_value += current_value
                has_real_trader = True

    return has_real_trader, has_simulated_trader, real_value, simulated_value


def _get_portfolios():
    simulated_portfolios = []
    real_portfolios = []

    for exchange_manager in interfaces.get_exchange_managers():
        if trading_api.is_trader_existing_and_enabled(exchange_manager):
            if trading_api.is_trader_simulated(exchange_manager):
                simulated_portfolios.append(trading_api.get_portfolio(exchange_manager))
            else:
                real_portfolios.append(trading_api.get_portfolio(exchange_manager))
    return real_portfolios, simulated_portfolios


def _merge_portfolios(base_portfolio, to_merge_portfolio):
    for currency, asset in to_merge_portfolio.items():
        if currency not in base_portfolio:
            base_portfolio[currency] = {
                constants.PORTFOLIO_AVAILABLE: decimal.Decimal(0),
                constants.PORTFOLIO_TOTAL: decimal.Decimal(0)
            }

        base_portfolio[currency][constants.PORTFOLIO_AVAILABLE] += asset.available
        base_portfolio[currency][constants.PORTFOLIO_TOTAL] += asset.total


def get_global_portfolio_currencies_amounts():
    real_portfolios, simulated_portfolios = _get_portfolios()
    real_global_portfolio = {}
    simulated_global_portfolio = {}

    for portfolio in simulated_portfolios:
        _merge_portfolios(simulated_global_portfolio, portfolio)

    for portfolio in real_portfolios:
        _merge_portfolios(real_global_portfolio, portfolio)

    return real_global_portfolio, simulated_global_portfolio


def get_global_portfolio_currencies_values() -> dict:
    return trading_api.get_global_portfolio_currencies_values(
        interfaces.get_exchange_managers()
        )
                

def trigger_portfolios_refresh():
    return interfaces.run_in_bot_main_loop(async_trigger_portfolios_refresh())


async def async_trigger_portfolios_refresh():
    at_least_one = False
    for exchange_manager in interfaces.get_exchange_managers():
        if trading_api.is_trader_existing_and_enabled(exchange_manager):
            at_least_one = True
            await trading_api.refresh_real_trader_portfolio(exchange_manager)

    if not at_least_one:
        raise RuntimeError("no real trader to update.")

File: packages/tentacles/Services/Interfaces/web_interface/static/js/components/portfolio.js
/*
 * Drakkar-Software OctoBot
 * Copyright (c) Drakkar-Software, All rights reserved.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3.0 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library.
 */

$(document).ready(function() {

    const createPortfolioChart = (element_id, title, update) => {
        const data = {};
        const element = $(`#${element_id}`);
        const max_medium_screen_legend_items = 50;
        const max_mobile_legend_items = 6;
        let at_least_one_value = false;
        let displayLegend = true;
        let graphHeight = element.attr("data-md-height");
        if(isMobileDisplay()){
            graphHeight = element.attr("data-sm-height");
        }
        element.attr("height", graphHeight);

        $(".symbol-holding").each(function (){
            const total_value = $(this).find(".total-value").text();
            if($.isNumeric(total_value)){
                data[$(this).find(".symbol").text()] = Number(total_value);
                if(Number(total_value) > 0 ){
                    at_least_one_value = true;
                }
            }
        });
        const dataLength = Object.keys(data).length;
        // display graph only if at least one value is available
        if(at_least_one_value && dataLength > 0 && element.length > 0){
            if(isMobileDisplay() && dataLength > max_mobile_legend_items){
                // legend is hiding the chart on smaller displays if too many elements are present
                displayLegend = false;
            }else if(dataLength > max_medium_screen_legend_items){
                // legend is hiding the chart on if too many elements are present
                displayLegend = false;
            }
            create_doughnut_chart(element[0], data, title, displayLegend, graphHeight, update);
        }else{
            element.addClass(hidden_class);
        }
    }

    const handle_portfolio_button = () => {
        const refreshButton = $("#refresh-portfolio");
        if(refreshButton){
            refreshButton.click(function () {
                const update_url = refreshButton.attr(update_url_attr);
                send_and_interpret_bot_update({}, update_url, null, generic_request_success_callback, generic_request_failure_callback);
            });
        }
    }

    const start_periodic_refresh = () => {
        setInterval(function() {
            $("#portfolio-display").load(location.href + " #portfolio-display", function (){
                update_display(true, true);
            });
        }, portfolio_update_interval);
    }

    const displayPortfolioTable = () => {
        handle_rounded_numbers_display();
        ordersDataTable = $('#holdings-table').DataTable({
            "paging": false,
            "bDestroy": true,
            "order": [[ 2, "desc" ]],
            "searching": $("tr.symbol-holding").length > 10,
        });
    }

    const displayPortfolioContent = (referenceMarket, update) => {
        displayPortfolioTable();
        const chartTitle = `Assets value (${referenceMarket})`;
        createPortfolioChart("portfolio_doughnutChart", chartTitle, update);
        handleButtons();
    }

    const update_display = (withImages, update) => {
        const referenceMarket = $("#portfoliosCard").attr("reference_market");
        displayPortfolioContent(referenceMarket, update);
        if(withImages){
            handleDefaultImages();
        }
    }

    let firstLoad = true;
    const handleClearButton = () => {
        $("#clear-portfolio-history-button").on("click", (event) => {
            if (confirm("Clear portfolio history ?") === false) {
                return false;
            }
            const url = $(event.currentTarget).data("url")
            const success = (updated_data, update_url, dom_root_element, msg, status) => {
                // reload page on success
                location.reload();
            }
            send_and_interpret_bot_update(null, url, null, success, generic_request_failure_callback)
        })
    }

    const handleButtons = () => {
        handle_portfolio_button();
        handleClearButton()
    }
    update_display(false, false);
    if(firstLoad){
        start_periodic_refresh();
    }
});
